internal/controller/linodefirewall_controller_helpers.go

```diff
@@ -26,15 +35,81 @@ var (
 	errTooManyIPs = errors.New("too many IPs in this ACL, will exceed rules per firewall limit")
 )
 
+func findObjectsForAddressSet(logger logr.Logger, tracedClient client.Client) handler.MapFunc {
+	logger = logger.WithName("LinodeFirewallReconciler").WithName("findObjectsForAddressSet")
+	return func(ctx context.Context, obj client.Object) []ctrl.Request {
+		ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultMappingTimeout)
+		defer cancel()
+
+		// Get all Firewalls because we can't filter on arbitrary fields in the spec
+		firewalls := &infrav1alpha2.LinodeFirewallList{}
+		if err := tracedClient.List(ctx, firewalls, &client.ListOptions{}); err != nil {
+			switch {
+			case apierrors.IsNotFound(err) || firewalls == nil:
+				logger.Info("LinodeFirewall(s) not found for %s")
+
+				return nil
+			case err != nil:
+				logger.Error(err, "Failed to get LinodeFirewalls")
+
+				return nil
+			}
+		}
+
+		return buildRequests(firewalls.Items, obj)
+	}
+}
+
+// Constructs a unique list of requests for updating LinodeFirewalls that either reference the
+// AddressSet in the Inbound and/or OutboundRules
+func buildRequests(firewalls []infrav1alpha2.LinodeFirewall, obj client.Object) []reconcile.Request {
+	requestSet := make(map[reconcile.Request]struct{})
+	for _, item := range firewalls {
+		for _, inboundRule := range item.Spec.InboundRules {
+			for _, addrSetRef := range inboundRule.AddressSetRefs {
+				if addrSetRef.Name == obj.GetName() && addrSetRef.Namespace == obj.GetNamespace() {
+					requestSet[reconcile.Request{
+						NamespacedName: types.NamespacedName{
+							Name:      item.GetName(),
+							Namespace: item.GetNamespace(),
+						},
+					}] = struct{}{}
+				}
+			}
+		}
+		for _, outboundRule := range item.Spec.OutboundRules {
+			for _, addrSetRef := range outboundRule.AddressSetRefs {
+				if addrSetRef.Name == obj.GetName() && addrSetRef.Namespace == obj.GetNamespace() {
+					requestSet[reconcile.Request{
+						NamespacedName: types.NamespacedName{
+							Name:      item.GetName(),
+							Namespace: item.GetNamespace(),
+						},
+					}] = struct{}{}
+				}
+			}
+		}
+	}
+	requests := make([]reconcile.Request, 0, len(requestSet))
```

Comment:
I think go 1.23 has maps.Values that returns an iterator. You can combine it with slices.Collect() to get a slice.